### ISSUE TITLE
Investigate Image Compression Across Gitcoin for Performance Improvements 

### ISSUE THREAD

**Issue by mkosowsk** [gitcoinco/web#608]
Note: this issue is meant for beginners as a first contribution to Gitcoin. As such, if you have previously made contributions to the gitcoinco/web repo please do not work on this issue... that means you can look but please don't touch @thelostone-mc 👀 

From #592:

Images on Gitcoin could possibly be optimized by compressing the file size without sacrificing too much quality. A good place to start on optimizing images for web is the Images section of

https://medium.freecodecamp.org/a-beginners-guide-to-website-optimization-2185edca0b72

Research different techniques and apply them to the images in the gitcoinco/web repo. Comment on this issue showing decreases in file size as well as before and after images to show if there's significant loss in quality.

Use a tool such as Chrome's Network Analysis (https://developers.google.com/web/tools/chrome-devtools/network-performance/reference#summary) to screen shot performance improvements for different compression techniques.

After your pull request is merged in, we can see via Google PageSpeed the improvements that your changes brought about! Could make a nice bullet point for the resume :)

<img width="490" alt="mobile" src="https://user-images.githubusercontent.com/7970539/37317776-37e95b5e-263d-11e8-83e8-764659118a65.png">

<img width="491" alt="desktop" src="https://user-images.githubusercontent.com/7970539/37317780-3b2ca726-263d-11e8-822a-68e366e37ac7.png">

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 0.1 ETH (69.09 USD @ $690.86/ETH) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/issue/gitcoinco/web/608).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/issue/gitcoinco/web/608)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $4166.23 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by johan718** [comment]
This is getting interesting ;).

**Comment by gitcoinbot** [comment]
__Work has been started on the 0.1 ETH (70.05 USD @ $700.53/ETH) funding by__: 
 1. [@cassidypignatello](https://gitcoin.co/profile/cassidypignatello) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/gitcoinco/web/608)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $4154.53 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by cassidypignatello** [comment]
Just wanted to confirm that I'll be working on this and will comment with screenshots once it's ready!

**Comment by mkosowsk** [comment]
@cassidypignatello looking forward to this, should be able to get some nice performance improvements! 👍

**Comment by cassidypignatello** [comment]
@mkosowsk thanks! Just wanted to leave a comment here showing some of the changes in file size / performance improvements. I only included a few screenshots of the file size reduction for redundancy's sake, but please let me know if you'd like to see more of those.

# **Before and After File Size Reduction**

<img width="978" alt="image-reduction-1" src="https://user-images.githubusercontent.com/6751725/37438571-ceda4ef0-27c9-11e8-95a9-7203c36320d7.png">
<img width="967" alt="image-reduction-2" src="https://user-images.githubusercontent.com/6751725/37438573-d24462ec-27c9-11e8-928d-527420cce0a1.png">
<img width="973" alt="image-reduction-3" src="https://user-images.githubusercontent.com/6751725/37438590-e89b2dd2-27c9-11e8-9d5d-d0283e9b9a32.png">
<img width="965" alt="image-reduction-4" src="https://user-images.githubusercontent.com/6751725/37438591-e8ab05cc-27c9-11e8-81c9-dead67b1b72a.png">

# **Page Load Before Optimization**

<img width="1438" alt="before-optimization" src="https://user-images.githubusercontent.com/6751725/37438653-40b98324-27ca-11e8-82e1-ff1f9368818e.png">

# **Page Load After Optimization**

<img width="1437" alt="after-optimization" src="https://user-images.githubusercontent.com/6751725/37438660-4aeeffcc-27ca-11e8-8b18-f1d0fa8c226d.png">

By optimizing the images, I was able to reduce the page load from 8 seconds to just under 5 seconds, fire the `DOMContentLoaded` event in 1.5 seconds, and reduce the total time (and amount) of screenshots of the page load to be taken from 8 seconds to just over 2.

Also, I noticed in my local version of the repo that every image has a duplicate with a random string of letters and numbers next to it, like "built-by-oss.baa082849c79.png". Should I leave those alone or delete them? I'm guessing those are cached versions of the images.

This may require a separate issue, but I also noticed that some images had duplicates with both png and jpg. I'm not sure if this was intentional, but I saw that at least one of them, `footer.png` wasn't being referenced in the project so I thought it'd be worth mentioning.

**Comment by mkosowsk** [comment]
@cassidypignatello amazing work! What technique(s) did you employ to get your results?

Could you share 2-3 before-and-after image pairs that are representative of the change? @PixelantDesign might be a good to get a designer's eye to make sure that the compression doesn't go too far

I'm not sure about the random string png files or the regular duplicate png files. @owocki @mbeacom do we need to hang onto these or could we jettison them?

Great job! 👍 👍 👍

**Comment by thelostone-mc** [comment]
@cassidypignatello @mkosowsk  
Thoughts on swapping out a few of those images for svgs ? 
That oughta increase performance

**Comment by mkosowsk** [comment]
@thelostone-mc do we have SVG's handy? Creating SVG's probably outside scope of this ticket, no?

**Comment by thelostone-mc** [comment]
@owocki would you have the sketch files for these ? If yes exporting it as an svg is easy
Else we'll take care of it when we redesign the page :)

**Comment by olafghanizadeh** [comment]
@thelostone-mc Almost all the graphics on the current page is possible to display as SVG instead. I started on that some time back, but never finished.

**Comment by mbeacom** [comment]
It might be ideal to implement ongoing image compression via a Django package, post processor, custom script to handle image compression as a pre-deployment step versus compressing the images and committing them back to the repo.  This will provide us with a scalable solution versus a manual action step anytime an image changes.
There are multiple django and python packages that can achieve this.

**Comment by thelostone-mc** [comment]
@mbeacom Image Compression == Yay but SVG + Compression == Double YAY

**Comment by cassidypignatello** [comment]
Thanks @mkosowsk! I build a lot of rich media ads at my day job, so having to optimize image file size without reducing quality is something I deal with constantly. The main technique I used here was actually putting all the images into tinypng.com and letting that handle the compression for me. I also have some Photoshop techniques that I can use, but didn't have to do that here since I've found that TinyPNG almost always does a better job and is way faster than manually exporting each image individually. That being said, I totally agree with @mbeacom's point!

I'm on my work computer at the moment, but I'll be sure to send over some before and-after image pairs once I'm back home tonight!

**Comment by mkosowsk** [comment]
@cassidypignatello Awesome! 

@mbeacom For the purposes of this issue and bounty, I think adding the pre-deployment step to compress is outside of scope and wouldn't want to feature creep it in for the sake of paying out bounty to @cassidypignatello 

I think it could make sense to merge @cassidypignatello's changes in once PR is approved as it should deliver real value very soon to Gitcoin. We could hold onto non-compressed images and create an issue down the line for the pre-deployment step... sound good? What do you think @thelostone-mc ?

This shows the value of the GIPS strategy, the community could have iterated on a better initial ticket rather than yours truly flying off a bit half-cocked 😂

**Comment by mbeacom** [comment]
@mkosowsk If that's the case, you might have been able to just use the assets that are pregenerated for you via the Google pagespeed results:  https://developers.google.com/speed/pagespeed/insights/optimizeContents?url=https%3A%2F%2Fgitcoin.co%2F&strategy=desktop

@cassidypignatello What are you using to compress the images?

But I'm fine with that...  Ultimately, I rather maintain a directory of the original images and a separate source of compressed images as the output dir for the automation step. A simple CLI script could achieve the compression and output to a "build" dir.

**Comment by olafghanizadeh** [comment]
@mbeacom @mkosowsk @cassidypignatello 
My suggestion is to do like this, implement imagemin (https://www.npmjs.com/package/imagemin-webpack-plugin) as a part of the asset build step. This plugin can compress png, jpg, svg (using SVGO algo). I feel like it would be beneficial to move most of the frontend asset processing into webpack.

**Comment by olafghanizadeh** [comment]
A lot of the icons used on Gitcoin is also very small in filesize, so ideally we would make use of SVG sprites to lower the amount of HTTP requests.

**Comment by mkosowsk** [comment]
@mbeacom When I hit that link I get this error message:

> The referrer https://github.com/gitcoinco/web/issues/608 does not match the referrer restrictions configured on your API key. Please use the API Console to update your key restrictions.

Looks like he's using TinyPNG :D

So for deliverables for this issue @cassidypignatello would create an output dir that he would populate with his compressed images? Then a bit further on down the line we could use imagemin as part of the asset build step that would output compressed images to this same folder?

Of course if Cassidy is comfortable with hooking up imagemin that would be up to their discretion, but I don't want to feature creep someone out of the ticket as it was originally written... I've had this happen to me too many times at work 😂

**Comment by thelostone-mc** [comment]
![creep](https://cdn-images-1.medium.com/max/1600/0*CIFEK0FxxIKqSuC7.gif)

**Comment by mbeacom** [comment]
@mkosowsk I agree regarding feature creep, but I don't think they should be exclusive.  You can run a script over the directory and automatically generate the compressed assets.  

If my URL doesn't work, you can look at your pagespeed insights report, scroll to the bottom, and it's the last link.

**Comment by cassidypignatello** [comment]
@mkosowsk @mbeacom I'm comfortable with trying to hook up imagemin. If we're going to include it in this issue then I'll hold off on uploading the before and after image pairs since it sounds like it would make more sense to just start again from scratch and recompress everything via the script. I'll get started on that unless there are any other concerns!

**Comment by mkosowsk** [comment]
Sounds good to me!

**Comment by mkosowsk** [comment]
@cassidypignatello Spoke with @mbeacom and @owocki and we think the TinyPNG approach is good for a first pass and for completing this bounty :)

Please open a PR with that approach at your earliest convenience 👍

**Comment by mkosowsk** [comment]
Will be interesting to check out PageSpeed Insights once changes are live!

**Comment by mbeacom** [comment]
<img width="394" alt="screenshot 2018-03-26 19 35 15" src="https://user-images.githubusercontent.com/7315957/37938626-d897d540-312c-11e8-95af-c6be9f3dbbbc.png">

<img width="372" alt="screenshot 2018-03-26 19 35 24" src="https://user-images.githubusercontent.com/7315957/37938640-e2caf77c-312c-11e8-9b4e-9d8d112b2077.png">

https://webpagetest.org/result/180326_KQ_4b50fee55bfaf252843c2d1f172ff439/
https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fgitcoin.co

**Comment by cassidypignatello** [comment]
Thanks @mkosowsk! As promised, here's a few images pair so everyone can check out the quality difference:

**Comment by cassidypignatello** [comment]
Thanks @mkosowsk! As promised, here's a few of the before and after image pairs so everyone can check out the quality difference:

grow_open_source.png (before) 1.3MB
![grow_open_source 5129c4417174](https://user-images.githubusercontent.com/6751725/37938723-45d34e1e-312d-11e8-8646-786981317cb2.png)

grow_open_source.png (after) 562KB
![grow_open_source](https://user-images.githubusercontent.com/6751725/37938750-6768f2fe-312d-11e8-975c-5f22868317c1.png)

divine.png (before) 179KB
![divine ad87d097c806](https://user-images.githubusercontent.com/6751725/37938765-7724fd64-312d-11e8-9ce0-5b9f6fe9c02e.png)

divine.png (after) 91KB
![divine](https://user-images.githubusercontent.com/6751725/37938785-8cf35f96-312d-11e8-9b34-ad9ebe5883d9.png)

dogfood.jpg (before) 92KB
![dogfood ac16b44c51d9](https://user-images.githubusercontent.com/6751725/37938831-bb207502-312d-11e8-8414-b30c11d1f999.jpg)

dogfood.jpg (after) 58KB
![dogfood](https://user-images.githubusercontent.com/6751725/37938836-c18a75d2-312d-11e8-8169-d25368da31fc.jpg)

**Comment by gitcoinbot** [comment]
__Work for 0.1 ETH (47.47 USD @ $474.72/ETH) has been submitted by__: 
 1. [@cassidypignatello](https://gitcoin.co/profile/cassidypignatello) 

 Submitters, please leave a comment to let the funder (@owocki) (and the other parties involved) that you've submitted you work.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/gitcoinco/web/608)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2759.07 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by mkosowsk** [comment]
From Medium to Good and Low to Medium... awesome!

**Comment by gitcoinbot** [comment]
__The funding of 0.1 ETH (45.69 USD @ $456.87/ETH) attached to this issue has been approved & issued to @cassidypignatello.__  

 * Learn more at [on the gitcoin issue page](https://gitcoin.co/issue/gitcoinco/web/608)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2509.13 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by weberV** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This issue now has a funding of 0.1 ETH (69.21 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/gitcoinco/web/608/135).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $1,588.41 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
